transformer_engine/jax/cpp_extensions.py

```diff
@@ -2662,6 +2664,445 @@ def cross_fused_attn_bwd(q: jnp.ndarray, kv: jnp.ndarray, bias: jnp.ndarray,
                                                            is_training=is_training)
 
 
+class FusedAttnFwdPrimitive(BasePrimitive):
+    """
+    Fused Attention Forward Primitive
+    Query, key, value are seperated tensors
+    """
+    name = "te_fused_attn_forward"
+    multiple_results = True
+    impl_static_args = (7, 8, 9, 10, 11)
+    inner_primitive = None
+    outer_primitive = None
+
+    @staticmethod
+    def abstract(q_aval, k_aval, v_aval, bias_aval, q_seqlen_or_cu_seqlen_aval,
+                 kv_seqlen_or_cu_seqlen_aval, seed_aval, *, attn_bias_type, attn_mask_type,
+                 scaling_factor, dropout_probability, is_training):
+        """
+        Fused attention fwd abstract
+        """
+        q_dtype = dtypes.canonicalize_dtype(q_aval.dtype)
+        k_dtype = dtypes.canonicalize_dtype(k_aval.dtype)
+        v_dtype = dtypes.canonicalize_dtype(v_aval.dtype)
+        bias_dtype = dtypes.canonicalize_dtype(bias_aval.dtype)
+        assert q_dtype == k_dtype == v_dtype == bias_dtype
+        assert q_seqlen_or_cu_seqlen_aval.dtype == kv_seqlen_or_cu_seqlen_aval.dtype
+
+        *q_batch_shape, q_max_seqlen, num_heads, q_head_dim = q_aval.shape
+        *kv_batch_shape, kv_max_seqlen, num_gqa_groups, kv_head_dim = k_aval.shape
+        assert q_batch_shape == kv_batch_shape
+        assert q_head_dim == kv_head_dim
+        assert k_aval.shape == v_aval.shape
+        out_aval = q_aval.update(shape=q_aval.shape, dtype=q_dtype)
+
+        # backend determines the softmax buffer shape/dtype
+        backend = FusedAttnHelper(q_dtype, k_dtype, NVTE_QKV_Layout.NVTE_BSHD_BS2HD, attn_bias_type,
```

Comment:
Should the QKV layout for this call be `NVTE_QKV_Layout.NVTE_BSHD_BSHD_BSHD`?

Reply:
Yes, thanks for catching it.